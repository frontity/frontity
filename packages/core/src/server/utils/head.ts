import { HelmetData } from "react-helmet";
import { HeadTags } from "../../../types";

export default (helmet: HelmetData): HeadTags => {
  return {
    base: helmet.base.toString(),
    bodyAttributes: helmet.bodyAttributes.toString(),
    htmlAttributes: helmet.htmlAttributes.toString(),
    link: helmet.link.toString(),
    meta: helmet.meta.toString(),
    noscript: helmet.noscript.toString(),
    script: helmet.script.toString(),
    style: helmet.style.toString(),
    title: helmet.title.toString()
  };
};
