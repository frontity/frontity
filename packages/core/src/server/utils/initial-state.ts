import deepmerge from "deepmerge";
import { NormalizedSettings } from "@frontity/file-settings";
import { Package } from "@frontity/types";

export default ({
  settings,
  url
}: {
  settings: NormalizedSettings;
  url: URL;
}) => {
  let state: Package["state"] = {
    frontity: {
      name: settings.name,
      mode: settings.mode,
      platform: "server",
      initial: {
        path: (/^(.*)page\/\d+\/?$/.exec(url.pathname) || [
          null,
          url.pathname
        ])[1],
        page: parseInt(
          (/^.*page\/(\d+)\/?$/.exec(url.pathname) || [null, "1"])[1],
          10
        )
      },
      packages: settings.packages.map(pkg => pkg.name)
    }
  };

  // Merge the initial state with the general state of settings.
  state = deepmerge(settings.state, state, {
    clone: false
  });

  // Merge the state with each package state, in order.
  settings.packages.forEach(pkg => {
    state = deepmerge(state, pkg.state, {
      clone: false
    });
  });

  return state;
};
