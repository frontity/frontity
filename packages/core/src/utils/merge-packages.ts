import { Package } from "@frontity/types";
import deepmerge from "deepmerge";
import getVariable from "./get-variable";

/**
 * A package function.
 *
 * @returns The package settings.
 */
type PackageFunction = () => Package;

/**
 *
 * Overwrites the merge config for deep merge.
 *
 * @param _ - The target array which is ignored.
 * @param sourceArray - The Array to be returned.
 *
 * @returns The source array.
 */
const overwriteMerge: deepmerge.Options["arrayMerge"] = (_, sourceArray) => {
  return sourceArray;
};

/**
 * Type for parameters of the {@link mergePackages} function.
 */
interface MergePackages {
  /**
   * The packages interface to be used.
   */
  packages: {
    [name: string]: Package | PackageFunction;
  };

  /**
   * The state entry.
   */
  state: Package["state"];

  /**
   * The flag for overwritting the arrays.
   */
  overwriteArrays?: boolean;
}

/**
 * Merge all packages together in a single config that can be passed
 * to createStore.
 *
 * @param params - Defined in {@link MergePackages}.
 *
 * @returns A Frontity package that can be passed to the {@link createStore} function.
 */
const mergePackages = ({
  packages,
  state,
  overwriteArrays = false,
}: MergePackages): Package => {
  let config: Package = {
    roots: {},
    state: {},
    actions: {},
    libraries: {
      frontity: {},
    },
    server: {},
  };
  state.frontity.packages.forEach((name) => {
    const variable = getVariable(name, state.frontity.mode);
    const module = packages[variable];

    // If module exist extend the base config.
    if (module) {
      const pkg = typeof module === "function" ? module() : module;
      config = deepmerge(config, pkg, {
        clone: false,
      });
    }
  });
  // Save debug, which is the only value we want to retain.
  const debug = !!config.state?.frontity?.debug;
  // Merge the state from the packages with the initial state.
  config.state = deepmerge(config.state, state, {
    clone: true,
    arrayMerge: overwriteArrays ? overwriteMerge : undefined,
  });
  // Delete the name.
  delete config.name;
  // Restore the debug flag if we are not in production.
  if (process.env.NODE_ENV !== "production")
    config.state.frontity.debug = debug;

  return config;
};

export default mergePackages;
