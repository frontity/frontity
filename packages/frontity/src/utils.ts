import chalk from "chalk";
import { readdir as readDir } from "fs-extra";

export const isPackageNameValid = (name: string): boolean => {
  if (name.length >= 214) return false;
  if ("._".includes(name[0])) return false;
  const specialCharacters = /[/:*?"<>|]/;
  if (specialCharacters.test(name)) return false;
  const nameConventionMatch = /^(?:@[a-z0-9_.-]+\/)?[a-z0-9_.-]+$/;
  return nameConventionMatch.test(name);
};

export const isThemeNameValid = (name: string): boolean => {
  const nameConventionMatch = /^(?:@[\w-]+\/)?[\w-]+$/;
  return nameConventionMatch.test(name);
};

export const errorLogger = (error: Error, message?: string) => {
  console.error(
    chalk.bold.red("\nError: ") +
      chalk.red(error.message) +
      "\n\n" +
      (message || "") +
      `If you need help please visit ${chalk.underline.magenta(
        "https://community.frontity.org/"
      )}.\n`
  );
  process.exit(1);
};

export const isFrontityProjectRoot = async (path: string) => {
  const dirContent = await readDir(path);
  return dirContent.some(content =>
    /^frontity\.settings\.(js|ts)$/i.test(content)
  );
};
