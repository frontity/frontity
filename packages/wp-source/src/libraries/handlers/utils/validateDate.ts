import { ServerError } from "@frontity/source";

/**
 *
 * Check if a given year is a leap year.
 *
 * @param year - Year as integer.
 *
 * @returns If the year is a leap year.
 */
function isLeapYear(year: number) {
  return (year % 4 == 0 && year % 100 != 0) || year % 400 == 0;
}

/**
 * Check that the given combination of year, month & day is valid.
 *
 * @param year - Year as integer.
 * @param month - Month of the year as integer (should start with 1).
 * @param day - Day of the month as integer (should start with 1).
 *
 */
function validateDate(year: number, month: number, day: number) {
  // This is just an arbitrary sanity check.
  // We ll generously assume that frontity will continue to exist for a hundred years ðŸ˜…
  if (year < 1900 || year > 2100) {
    throw new ServerError(`${year} is a wrong year number`, 404);
  }

  const monthDays = [31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  if (isLeapYear(year)) {
    monthDays[1] = 29;
  }

  if (typeof month === "number" && (month < 1 || month > 12)) {
    throw new ServerError(`${month} is a wrong month number`, 404);
  }
  if (typeof day === "number" && (day < 1 || day > monthDays[month - 1])) {
    throw new ServerError(`${day} is a wrong day number`, 404);
  }
}

export default validateDate;
