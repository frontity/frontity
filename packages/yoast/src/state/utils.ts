import { YoastSocialDefaults, YoastMeta } from "../../types";
import { Author, Taxonomy, PostType, Attachment } from "@frontity/source/";

export type Entity = (Author | Taxonomy | PostType | Attachment) & {
  yoast_meta: YoastMeta;
};

export const getEntity = ({ state, data }): Entity => {
  if (data.isPostType) return state.source[data.type][data.id];
  if (data.isTaxonomy) return state.source[data.taxonomy][data.id];
  if (data.isAuthor) return state.source.author[data.id];
  return null;
};

export const appendPage = ({
  title,
  page,
  totalPages
}: {
  title: string;
  page: number;
  totalPages: number;
}) => (page > 1 ? `${title} - Page ${page} of ${totalPages}` : title);

export const getSocialDefaults = ({ state }): YoastSocialDefaults => {
  const data = state.source.get(state.router.link);

  let type: string;
  let id: number;

  if (data.isArchive) ({ type, id } = data.items[0]);
  if (data.isPostType) ({ type, id } = data);

  if (type && id)
    return state.source[type][id].yoast_meta.yoast_wpseo_social_defaults;

  return {};
};

export const getPathname = route => {
  const [, pathname] = /^(?:(?:[^:/?#]+):)?(?:\/\/(?:[^/?#]*))?([^?#]*)/.exec(
    route
  );
  return pathname;
};

export const getPage = route => {
  const [, page] = /^(?:.*)page\/(\d+)\/?(\?.*)?$/.exec(getPathname(route)) || [
    null,
    "1"
  ];
  return parseInt(page, 10);
};

export const hasName = (
  entity: Entity
): entity is (Author | Taxonomy) & {
  yoast_meta: YoastMeta;
} => "name" in entity;
