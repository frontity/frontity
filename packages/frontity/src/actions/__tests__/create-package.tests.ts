import createPackage from "../create-package";
import inquirer from "inquirer";
import * as utils from "../../utils";

jest.mock("../../functions/create-package");
jest.mock("inquirer");
jest.mock("events");
jest.mock("../../utils");

const mockedInquirer = inquirer as jest.Mocked<typeof inquirer>;
const mockedUtils = utils as jest.Mocked<typeof utils>;
const mockedExit = jest.spyOn(process, "exit");

beforeEach(() => {
  mockedInquirer.prompt.mockRestore();
  mockedUtils.isFrontityProjectRoot.mockRestore();
  mockedExit.mockRestore();
  mockedExit.mockImplementation(() => {
    throw new Error();
  });
});

describe("create-package", () => {
  test("should not ask for passed options", async () => {
    const name = "example-theme";
    const namespace = "theme";
    await createPackage(name, { namespace });
    expect(mockedInquirer.prompt).not.toHaveBeenCalled();
  });

  test("should ask for missing name", async () => {
    mockedInquirer.prompt.mockResolvedValueOnce({ name: "example-theme" });

    const name = undefined;
    const namespace = "theme";
    await createPackage(name, { namespace });
    expect(mockedInquirer.prompt).toHaveBeenCalled();
    expect(mockedInquirer.prompt.mock.calls).toMatchSnapshot();
  });

  test("should ask for missing namespace", async () => {
    mockedInquirer.prompt.mockResolvedValueOnce({ namespace: "theme" });

    const name = "example-theme";
    const namespace = undefined;
    await createPackage(name, { namespace });
    expect(mockedInquirer.prompt).toHaveBeenCalled();
    expect(mockedInquirer.prompt.mock.calls).toMatchSnapshot();
  });

  test.todo("should fail in a invalid directory");
  test.todo("should work in a valid directory");
  test.todo("should fail with an invalid name");
  test.todo("should work with a valid name");
  test.todo("should work with a valid name with scope");
  test.todo("should fail with an invalid namespace");
  test.todo("should work with a valid namespace");
});
