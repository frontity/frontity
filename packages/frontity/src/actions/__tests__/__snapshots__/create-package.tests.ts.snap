// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create-package should ask for missing name 1`] = `
Array [
  Array [
    Array [
      Object {
        "default": "my-frontity-package",
        "message": "Enter a name for the package:",
        "name": "name",
        "type": "input",
      },
    ],
  ],
]
`;

exports[`create-package should ask for missing namespace 1`] = `
Array [
  Array [
    Array [
      Object {
        "default": "theme",
        "message": "Enter the namespace of the package:",
        "name": "namespace",
        "type": "input",
      },
    ],
  ],
]
`;
